### DISCUSSION TITLE
How to get sequence coordinates from a click event?

### DISCUSSION THREAD

**Discussion by dlrice** [molstar/molstar#111] [Q&A] [answered]
How can I extract the sequence coordinates from a click event? I.e.
```
    this.plugin.behaviors.interaction.click.subscribe((event) => {
      if (event.current) {
        //event.current.some.path.to.residue.coordinate?
      }
    });
```
In the example below, a user clicks on the 50th amino acid (tyrosine in this case) and the coordinate 50 is displayed in the LociLabels component:

<img width="925" alt="Screenshot 2020-12-24 at 16 02 29" src="https://user-images.githubusercontent.com/4357962/103098370-6ac36000-4602-11eb-8fbd-b934a91ce980.png">

How can I extract this from the event object that is return from a PluginContext.behaviors.interaction.click event?

**Comment by dsehnal** [comment, marked as answer]
There is no "standard" utility to do this so you need to transform the ``current`` loci in the event manually. This is how it's done for the hover labels: https://github.com/molstar/molstar/blob/master/src/mol-theme/label.ts#L234

Let me know if you need more pointers.

**Reply by dlrice** [reply]
Apologies for the extended silence. Thanks very much for pointing me to that code. My attempt so far is:
```
    this.plugin.behaviors.interaction.click.subscribe((event) => {
      if (event) {
        const { loci } = event.current;
        const el = loci?.elements?.[0];
        if (el) {
          const auth_seq_id = el.unit.model.atomicHierarchy.residues.auth_seq_id.value(
            el.unit.residueIndex[0] // not sure which index to use here
          );
        }
      }
    });
```
Which I found by looking at the reference you provided and looking through the calls:
```
=> _atomicElementLabel
=> StructureProperties.residue.auth_seq_id(location);
=> auth_seq_id: p(l => !Unit.isAtomic(l.unit) ? notAtomic() : l.unit.model.atomicHierarchy.residues.auth_seq_id.value(l.unit.residueIndex[l.element])),
```

The only challenge I have now is to find the correct index of `l.unit.residueIndex` as `l.element:number` doesn't exist on `loci?.elements?.[0]`.

**Reply by arose** [reply]
try `el.unit.residueIndex[OrderedSet.start(el.elements)]`

**Reply by dlrice** [reply]
`el.elements` is undefined for me but I can do `el.unit.elements` however this produces the same index as I click on different amino acids:

```
    this.plugin.behaviors.interaction.click.subscribe((event) => {
      if (event) {
        const { loci } = event.current;
        const el = loci?.elements?.[0];
        if (el) {
          const i = OrderedSet.start(el.unit.elements);
          const authSeqId = el.unit.model.atomicHierarchy.residues.auth_seq_id.value(
            el.unit.residueIndex[i]
          );
          console.log(i, authSeqId); // => 0 1 regardless of which AA is clicked
        }
      }
    });
```

**Reply by dlrice** [reply]
Any additional suggestions or documentation to approach this?

**Reply by dsehnal** [reply]
Sorry Daniel for the late reply.

The canonical solution would be 

```ts
this.plugin.behaviors.interaction.click.subscribe((event) => {
    if (StructureElement.Loci.is(event.current.loci)) {

        const loc = StructureElement.Location.create();
        StructureElement.Loci.getFirstLocation(event.current.loci, loc);
        // or loc = StructureElement.Loci.getFirstLocation(event.current.loci) which is ok to use if you dont do many sequential queries

        console.log(StructureProperties.residue.auth_seq_id(loc));
    }
});
```

-------------------

The mistake in your code is in the line ``const i = OrderedSet.start(el.unit.elements);``. It should be ``OrderedSet.start(el.indices)``.

**Reply by dlrice** [reply]
Thanks so much!
